### PR TITLE
Allow python modules to be created for existing BESS modules.

### PR DESCRIPTION
**Pull request by justinemarie**
This allows you to instantiate a module in python for a BESS module that already exists; just invoke the module by name and declare _do_not_create in the constructor. An example is in conf/samples/reattachment.bess.

### CODE REVIEW

**Comment by justinemarie** [comment]
Wait one more thing I need to add (check to make sure module exists).

**Comment by justinemarie** [comment]
Okay put the right check in now :-)

File: bessctl/module.py [hunk 1]
```diff
@@ -22,10 +22,21 @@ def __init__(self, **kwargs):
...
+            #create an object
+            ret = self.bess.create_module(self.__class__.__name__, name,
                                       self.choose_arg(None, kwargs))
-
-        self.name = ret.name
+            self.name = ret.name
+        else:
+            #bind to a pre-existing object, check if it's real
+            assert self.name != None
+            all_modules = self.bess.list_modules().modules
```
Comment (sangjinhan) [inline bessctl/module.py]: `list_modules()` enumerates all modules, thus potentially expensive if there are a lot. Instead you can use (untested):
```
info = self.bess.get_module_info(name)
check if self.__class__.__name__ agrees with info.mclass
```

File: bessctl/module.py [hunk 2]
```diff
@@ -22,10 +22,21 @@ def __init__(self, **kwargs):
...
+        if(not _do_not_create):
+            #create an object
+            ret = self.bess.create_module(self.__class__.__name__, name,
                                       self.choose_arg(None, kwargs))
-
-        self.name = ret.name
+            self.name = ret.name
+        else:
+            #bind to a pre-existing object, check if it's real
+            assert self.name != None
```
Comment (sangjinhan) [inline bessctl/module.py]: `name`, not `self.name`

File: sanity_check.sh [hunk 3]
```diff
@@ -13,7 +13,8 @@ TESTS='exactmatch.bess
 	l2_forward.bess
 	multicore.bess
 	queue.bess
-	reattachment.bess
+  reattachment.bess
+  reattachment.bess
```
Comment (sangjinhan) [inline sanity_check.sh]: Why twice?

Also, the file has mixed use of tabs and spaces.

File: bessctl/module.py (hunk 2 above)
Comment (justinemarie) [inline bessctl/module.py]: fixed

File: bessctl/module.py (hunk 1 above)
Comment (justinemarie) [inline bessctl/module.py]: fixed

File: sanity_check.sh (hunk 3 above)
Comment (justinemarie) [inline sanity_check.sh]: Copy/paste error :-) Fixed.

**Comment by justinemarie** [comment]
This is ready for another lookthrough and/or commit as soon as it finishes in Travis.

File: bessctl/module.py [hunk 4]
```diff
@@ -22,10 +22,18 @@ def __init__(self, **kwargs):
         else:
             name = None
 
-        ret = self.bess.create_module(self.__class__.__name__, name,
+        if(not _do_not_create):
```
Comment (sangjinhan) [inline bessctl/module.py]: no parenthesis here

File: bessctl/module.py [hunk 5]
```diff
@@ -22,10 +22,18 @@ def __init__(self, **kwargs):
...
+            ret = self.bess.create_module(self.__class__.__name__, name,
                                       self.choose_arg(None, kwargs))
-
-        self.name = ret.name
+            self.name = ret.name
+        else:
+            #bind to a pre-existing object, check if it's real
+            assert name != None
+            info = self.bess.get_module_info(name)
+            assert info != None
```
Comment (sangjinhan) [inline bessctl/module.py]: Checking `info` here doesn't do anything. It won't return `None` at all. Just either success or exception.

File: sanity_check.sh [hunk 6]
```diff
@@ -29,7 +30,7 @@ function fail
 }
 
 echo "This script runs a collection of BESS sample scripts and makes sure \
-nothing blows up. Sit back and relax."
+  nothing blows up. Sit back and relax."
```
Comment (sangjinhan) [inline sanity_check.sh]: You have three spaces between `sure` and `nothing`.

File: bessctl/conf/samples/reattachment.bess [hunk 7]
```diff
@@ -0,0 +1,50 @@
...
+
+# locally emulate two input ports
+Source() -> SetMetadata(attrs=[{'name': 'in_port', 'size': 2, 'value_int': 1}]) -> Rewrite(templates=pkts) -> wm
+
+wm:0 -> Sink()
+wm:1 -> Sink()
+wm:2 -> Sink()
+wm.set_default_gate(gate=2)
+
+# 172.16.100.1/32, port 80, in_port 1
```
Comment (sangjinhan) [inline bessctl/conf/samples/reattachment.bess]: As it is derived from the `wildcardmatch.bess` example, we have quite redundancy which is not quite relevant to the point of this test script. I would suggest keeping only `wm` and nothing else to make the purpose of the test clearer.

File: sanity_check.sh [hunk 8]
```diff
@@ -5,21 +5,22 @@ SCRIPTS='./bessctl/conf/samples'
...
-	wildcardmatch.bess'
+flowgen.bess
+generic_encap.bess
+hash_lb.bess
+igate.bess
+iplookup.bess
+l2_forward.bess
+multicore.bess
+queue.bess
+reattachment.bess
```
Comment (sangjinhan) [inline sanity_check.sh]: Perhaps a bit off the topic, but we should be careful to add files here. This sanity test takes about half of the total Travis build time, being the most time consuming step (and keeps growing).

Given that what this script does is mostly identical to `wildcardmatch.bess` below, how about not adding this file, until we derive a solution for the sanity check to shorten its test time?

File: bessctl/module.py [hunk 9]
```diff
@@ -22,10 +22,18 @@ def __init__(self, **kwargs):
...
-        ret = self.bess.create_module(self.__class__.__name__, name,
+        if(not _do_not_create):
+            #create an object
+            ret = self.bess.create_module(self.__class__.__name__, name,
                                       self.choose_arg(None, kwargs))
-
-        self.name = ret.name
+            self.name = ret.name
+        else:
+            #bind to a pre-existing object, check if it's real
```
Comment (sangjinhan) [inline bessctl/module.py]: Space

File: bessctl/module.py [hunk 10]
```diff
@@ -22,10 +22,18 @@ def __init__(self, **kwargs):
         else:
             name = None
 
-        ret = self.bess.create_module(self.__class__.__name__, name,
+        if(not _do_not_create):
+            #create an object
```
Comment (sangjinhan) [inline bessctl/module.py]: Space

File: sanity_check.sh (hunk 8 above)
Comment (justinemarie) [inline sanity_check.sh]: Good point. I will do this. Why don't I just put the reattachment bit inside wildcardmatch.bess?

Comment (sangjinhan) [inline sanity_check.sh]: That should work too!

File: bessctl/module.py (hunk 4 above)
Comment (justinemarie) [inline bessctl/module.py]: Fixed (why?)

File: bessctl/module.py (hunk 9 above)
Comment (justinemarie) [inline bessctl/module.py]: What does this mean?

File: bessctl/module.py (hunk 5 above)
Comment (justinemarie) [inline bessctl/module.py]: Fixed

File: sanity_check.sh (hunk 6 above)
Comment (justinemarie) [inline sanity_check.sh]: fixed

**Comment by justinemarie** [comment]
*bump*

**Comment by melvinw** [comment]
Is `conf/samples/reattachment.bess` supposed to be part of this PR? I can't find it.

**Comment by justinemarie** [comment]
No, it has been removed.

File: bessctl/module.py [hunk 11]
```diff
@@ -22,10 +22,17 @@ def __init__(self, **kwargs):
...
-
-        self.name = ret.name
+        if not _do_not_create:
+            # create an object
+            ret = self.bess.create_module(self.__class__.__name__, name,
+                                          self.choose_arg(None, kwargs))
+            self.name = ret.name
+        else:
+            # bind to a pre-existing object, check if it's real
+            assert name != None
```
Comment (melvinw) [inline bessctl/module.py]: Could you add an error message to this assertion (e.g. https://github.com/NetSys/bess/blob/master/bessctl/commands.py#L92)

File: bessctl/module.py [hunk 12]
```diff
@@ -22,10 +22,17 @@ def __init__(self, **kwargs):
...
+            info = self.bess.get_module_info(name)
+            assert self.mclass == info.mclass
```
Comment (melvinw) [inline bessctl/module.py]: Same here

File: bessctl/module.py (hunk 11 above)
Comment (justinemarie) [inline bessctl/module.py]: Fixed.

File: bessctl/module.py (hunk 12 above)
Comment (justinemarie) [inline bessctl/module.py]: Fixed.

**Comment by melvinw** [comment]
Thanks!
cat

**Pull request merged**